diff --git a/slurm/slurm.h b/slurm/slurm.h
index 655ff9f510..0c6cd24e2a 100644
--- a/slurm/slurm.h
+++ b/slurm/slurm.h
@@ -2930,6 +2930,8 @@ typedef struct reservation_name_msg {
  */
 typedef struct {
 	time_t last_update;	/* last update time of the build parameters */
+	char *sunk_node_dashboard_url; /* sunk node grafana dashboard url */
+	char *sunk_job_dashboard_url; /* sunk job grafana dashboard url */
 	char *accounting_storage_tres; /* list of tres */
 	uint16_t accounting_storage_enforce; /* job requires valid association:
 					      * user/account/partition/cluster */
diff --git a/src/api/config_info.c b/src/api/config_info.c
index 206c015d6b..7456cd2cd8 100644
--- a/src/api/config_info.c
+++ b/src/api/config_info.c
@@ -741,6 +741,9 @@ extern void *slurm_ctl_conf_2_key_pairs(slurm_conf_t *conf)
 		xfree(key);
 	}

+	add_key_pair(ret_list, "SUNKNodeDashboardURL", "%s", conf->sunk_node_dashboard_url);
+	add_key_pair(ret_list, "SUNKJobDashboardURL", "%s", conf->sunk_job_dashboard_url);
+
 	if (xstrcmp(conf->priority_type, "priority/basic")) {
 		add_key_pair(ret_list, "FairShareDampeningFactor", "%u",
 			     conf->fs_dampening_factor);
diff --git a/src/api/node_info.c b/src/api/node_info.c
index 42e20c2f8f..cf41d304ad 100644
--- a/src/api/node_info.c
+++ b/src/api/node_info.c
@@ -461,6 +461,15 @@ char *slurm_sprint_node_table(node_info_t *node_ptr, int one_liner)
 		xstrfmtcat(out, "ReservationName=%s", node_ptr->resv_name);
 	}

+	/****** Line (optional) ******/
+	if (slurm_conf.sunk_node_dashboard_url && node_ptr->comment) {
+		xstrcat(out, line_end);
+		// SUNK slurm node comment always contains the node name
+		xstrfmtcat(out, "SUNKNodeDashboardURL=%s&var-node=%s",
+			slurm_conf.sunk_node_dashboard_url,
+			node_ptr->comment);
+	}
+
 	if (one_liner)
 		xstrcat(out, "\n");
 	else
diff --git a/src/common/read_config.c b/src/common/read_config.c
index 5c58f65a56..a3445d81a9 100644
--- a/src/common/read_config.c
+++ b/src/common/read_config.c
@@ -316,6 +316,8 @@ s_p_options_t slurm_conf_options[] = {
 	{"LaunchType", S_P_STRING},
 	{"Licenses", S_P_STRING},
 	{"LogTimeFormat", S_P_STRING},
+	{"SUNKNodeDashboardURL", S_P_STRING},
+	{"SUNKJobDashboardURL", S_P_STRING},
 	{"MailDomain", S_P_STRING},
 	{"MailProg", S_P_STRING},
 	{"MaxArraySize", S_P_UINT32},
@@ -3013,6 +3015,8 @@ extern void free_slurm_conf(slurm_conf_t *ctl_conf_ptr, bool purge_node_hash)
 	xfree (ctl_conf_ptr->slurmd_spooldir);
 	xfree (ctl_conf_ptr->slurmd_user_name);
 	xfree (ctl_conf_ptr->srun_epilog);
+	xfree (ctl_conf_ptr->sunk_node_dashboard_url);
+	xfree (ctl_conf_ptr->sunk_job_dashboard_url);
 	xfree (ctl_conf_ptr->srun_port_range);
 	xfree (ctl_conf_ptr->srun_prolog);
 	xfree (ctl_conf_ptr->state_save_location);
@@ -3206,6 +3210,8 @@ void init_slurm_conf(slurm_conf_t *ctl_conf_ptr)
 	xfree (ctl_conf_ptr->state_save_location);
 	xfree (ctl_conf_ptr->suspend_exc_nodes);
 	xfree (ctl_conf_ptr->suspend_exc_parts);
+	xfree (ctl_conf_ptr->sunk_node_dashboard_url);
+	xfree (ctl_conf_ptr->sunk_job_dashboard_url);
 	xfree (ctl_conf_ptr->suspend_exc_states);
 	xfree (ctl_conf_ptr->suspend_program);
 	ctl_conf_ptr->suspend_rate		= NO_VAL16;
@@ -5501,6 +5507,9 @@ static int _validate_and_set_defaults(slurm_conf_t *conf,
 	(void) s_p_get_string(&conf->task_epilog, "TaskEpilog", hashtbl);
 	(void) s_p_get_string(&conf->task_prolog, "TaskProlog", hashtbl);

+	(void) s_p_get_string(&conf->sunk_node_dashboard_url, "SUNKNodeDashboardURL", hashtbl);
+	(void) s_p_get_string(&conf->sunk_job_dashboard_url, "SUNKJobDashboardURL", hashtbl);
+
 	if (!s_p_get_uint16(&conf->tcp_timeout, "TCPTimeout", hashtbl))
 		conf->tcp_timeout = DEFAULT_TCP_TIMEOUT;

diff --git a/src/scontrol/info_job.c b/src/scontrol/info_job.c
index 2526f0627c..d7c5b231e6 100644
--- a/src/scontrol/info_job.c
+++ b/src/scontrol/info_job.c
@@ -909,6 +909,15 @@ static char *_sprint_job_info(job_info_t *job_ptr)
 		xstrfmtcat(out, "ResvPorts=%s", job_ptr->resv_ports);
 	}

+	/****** Line (optional) ******/
+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name) {
+		xstrcat(out, line_end);
+		xstrfmtcat(out, "SUNKJobDashboardURL=%s&var-slurm_cluster=%s&var-job_id=%d",
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			job_ptr->job_id);
+	}
+
 	xstrcat(out, line_end);

 	/****** END OF JOB RECORD ******/
diff --git a/src/slurmd/slurmd/req.c b/src/slurmd/slurmd/req.c
index 67f4572c5b..497cfc407e 100644
--- a/src/slurmd/slurmd/req.c
+++ b/src/slurmd/slurmd/req.c
@@ -2738,6 +2738,14 @@ static void _rpc_batch_job(slurm_msg_t *msg)

 	info("Launching batch job %u for UID %u", req->job_id, batch_uid);

+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name) {
+		info("Job %d: SUNK job dashboard: %s&var-slurm_cluster=%s&var-job_id=%d",
+			req->job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			req->job_id);
+	}
+
 	debug3("_rpc_batch_job: call to _forkexec_slurmstepd");
 	rc = _forkexec_slurmstepd(LAUNCH_BATCH_JOB, (void *)req, cli, batch_uid,
 				  req->job_id, SLURM_BATCH_SCRIPT,
@@ -5336,6 +5344,14 @@ _rpc_terminate_job(slurm_msg_t *msg)
 	int             nsteps = 0;
 	int		delay;

+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name && req->step_id.job_id) {
+		info("Job %d: SUNK job dashboard: %s&var-slurm_cluster=%s&var-job_id=%d",
+			req->step_id.job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			req->step_id.job_id);
+	}
+
 	debug("%s: uid = %u %ps", __func__, msg->auth_uid, &req->step_id);
 	/*
 	 * check that requesting user ID is the Slurm UID
diff --git a/src/srun/launch.c b/src/srun/launch.c
index 55e8a82c0f..e3b823e255 100644
--- a/src/srun/launch.c
+++ b/src/srun/launch.c
@@ -1520,6 +1520,14 @@ extern int launch_g_step_launch(srun_job_t *job, slurm_step_io_fds_t *cio_fds,
 		}
 	}

+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name && job->step_id.job_id) {
+		printf("Job %d: SUNK job dashboard: %s&var-slurm_cluster=%s&var-job_id=%d\n",
+			job->step_id.job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			job->step_id.job_id);
+	}
+
 	update_job_state(job, SRUN_JOB_STARTING);
 	if (slurm_step_launch_wait_start(job->step_ctx) == SLURM_SUCCESS) {
 		update_job_state(job, SRUN_JOB_RUNNING);
diff --git a/src/srun/srun_job.c b/src/srun/srun_job.c
index 98d19a56c2..6b8a1628db 100644
--- a/src/srun/srun_job.c
+++ b/src/srun/srun_job.c
@@ -1596,6 +1596,14 @@ extern void fini_srun(srun_job_t *job, bool got_alloc, uint32_t *global_rc)
 		*global_rc = 128 + WTERMSIG(*global_rc);

 	mpir_cleanup();
+
+	if (slurm_conf.sunk_job_dashboard_url && slurm_conf.cluster_name && job->step_id.job_id) {
+		printf("Job %d: SUNK job dashboard: %s&var-slurm_cluster=%s&var-job_id=%d\n",
+			job->step_id.job_id,
+			slurm_conf.sunk_job_dashboard_url,
+			slurm_conf.cluster_name,
+			job->step_id.job_id);
+	}
 }

 void
