diff --git a/src/slurmctld/read_config.c b/src/slurmctld/read_config.c
index c16868c2b4..467e7e4512 100644
--- a/src/slurmctld/read_config.c
+++ b/src/slurmctld/read_config.c
@@ -1584,18 +1584,6 @@ extern int read_slurm_conf(int recover)
 		goto end_it;
 	}

-	/*
-	 * Node reordering may be done by the topology plugin.
-	 * Reordering the table must be done before hashing the
-	 * nodes, and before any position-relative bitmaps are created.
-	 *
-	 * Sort the nodes read in from the slurm.conf first before restoring
-	 * the dynamic nodes from the state file to prevent dynamic nodes from
-	 * being sorted -- which can cause problems with heterogenous jobs and
-	 * the order of the sockets changing on startup.
-	 */
-	_sort_node_record_table_ptr();
-
 	/*
 	 * Load node state which includes dynamic nodes so that dynamic nodes
 	 * can be included in topology.
@@ -1614,6 +1602,13 @@ extern int read_slurm_conf(int recover)
 		(void) load_all_front_end_state(false);
 	}

+	/*
+	 * Node reordering may be done by the topology plugin.
+	 * Reordering the table must be done before hashing the
+	 * nodes, and before any position-relative bitmaps are created.
+	 */
+	_sort_node_record_table_ptr();
+
 	rehash_node();
 	topology_g_build_config();

