diff --git a/src/plugins/topology/block/block_record.c b/src/plugins/topology/block/block_record.c
index 050b4716de..18d85b255f 100644
--- a/src/plugins/topology/block/block_record.c
+++ b/src/plugins/topology/block/block_record.c
@@ -113,7 +113,7 @@ static int _read_topo_file(slurm_conf_block_t **ptr_array[])
 	debug("Reading the %s file", topo_conf);

 	conf_hashtbl = s_p_hashtbl_create(block_options);
-	if (s_p_parse_file(conf_hashtbl, NULL, topo_conf, false, NULL) ==
+	if (s_p_parse_file(conf_hashtbl, NULL, topo_conf, PARSE_FLAGS_IGNORE_NEW, NULL) ==
 	    SLURM_ERROR) {
 		s_p_hashtbl_destroy(conf_hashtbl);
 		fatal("something wrong with opening/reading %s: %m",
diff --git a/src/plugins/topology/tree/switch_record.c b/src/plugins/topology/tree/switch_record.c
index 478bb59cd2..b240aca808 100644
--- a/src/plugins/topology/tree/switch_record.c
+++ b/src/plugins/topology/tree/switch_record.c
@@ -181,7 +181,7 @@ static int  _read_topo_file(slurm_conf_switches_t **ptr_array[])
 	debug("Reading the %s file", topo_conf);

 	conf_hashtbl = s_p_hashtbl_create(switch_options);
-	if (s_p_parse_file(conf_hashtbl, NULL, topo_conf, 0, NULL) ==
+	if (s_p_parse_file(conf_hashtbl, NULL, topo_conf, PARSE_FLAGS_IGNORE_NEW, NULL) ==
 	    SLURM_ERROR) {
 		s_p_hashtbl_destroy(conf_hashtbl);
 		fatal("something wrong with opening/reading %s: %m",
