diff --git a/src/slurmd/common/xcpuinfo.c b/src/slurmd/common/xcpuinfo.c
index c970828918..28ec2663c0 100644
--- a/src/slurmd/common/xcpuinfo.c
+++ b/src/slurmd/common/xcpuinfo.c
@@ -1292,7 +1292,8 @@ extern char *xcpuinfo_get_cpuspec(void)
 	res_cpu_bitmap = bit_alloc(MAX_CPU_CNT);
 	bit_unfmt(res_core_bitmap, restricted_cpus_as_abs);

-	for (int core_off = 0; core_off < conf->cores; core_off++) {
+	int total_cores = conf->sockets * conf->cores;
+	for (int core_off = 0; core_off < total_cores; core_off++) {
 		if (!bit_test(res_core_bitmap, core_off))
 			continue;
 		for (int thread_off = 0; thread_off < conf->threads;
